plugins/redis/redisasync.cpp

```diff
@@ -0,0 +1,870 @@
+{
+    channel.set(_channel);
+    callback = subCB;
+}
+SubContainer::SubContainer(ICodeContext * ctx, const char * options, const char * _channel, redisCallbackFn * _callback, unsigned __int64 _database) : AsyncConnection(ctx, options,_database )
+{
+    channel.set(_channel);
+    callback = _callback;
+}
+AsyncConnection::AsyncConnection(ICodeContext * ctx, const char * _options, unsigned __int64 _database) : Connection(ctx, _options, _database)
+{
+    createAndAssertConnection(ctx);
+    //could log server stats here, however async connections are not cached and therefore book keeping of only doing so for new servers may not be worth it.
+}
+void AsyncConnection::selectDb(ICodeContext * ctx)
+{
+    if (database == 0)
+        return;
+    attachLibev();
+    VStringBuffer cmd("SELECT %llu", database);
+    assertRedisErr(redisAsyncCommand(context, selectCB, NULL, cmd.str()), "SELECT (lock) buffer write error");
+    handleLoop(EV_DEFAULT_ EV_TIMEOUT);
+}
+AsyncConnection::~AsyncConnection()
+{
+    if (context)
+    {
+        //redis can auto disconnect upon certain errors, disconnectCB is called to handle this and is automatically
+        //disconnected after this freeing the context
+        if (context->err == REDIS_OK)
+            redisAsyncDisconnect(context);
+    }
+}
+void AsyncConnection::createAndAssertConnection(ICodeContext * ctx)
+{
+    context = NULL;
+    context = redisAsyncConnect(master.str(), port);
+    assertConnection();
+    context->data = (void*)this;
+    assertRedisErr(redisAsyncSetConnectCallback(context, connectCB), "failed to set connect callback");
+    assertRedisErr(redisAsyncSetDisconnectCallback(context, disconnectCB), "failed to set disconnect callback");
+    selectDb(ctx);
+}
+void AsyncConnection::assertConnection()
+{
+    if (!context)
+        rtlFail(0, "Redis Plugin: async context mem alloc fail.");
+    assertContextErr(context);
+}
+void AsyncConnection::assertContextErr(const redisAsyncContext * context)
+{
+    if (context && context->err)
+    {
+        const AsyncConnection * connection = (const AsyncConnection*)context->data;
+        if (connection)
+        {
+            VStringBuffer msg("Redis Plugin : failed to create connection context for %s:%d - %s", connection->getMaster(), connection->getPort(), context->errstr);
+            rtlFail(0, msg.str());
+        }
+        {
+            VStringBuffer msg("Redis Plugin : failed to create connection context - %s", context->errstr);
+            rtlFail(0, msg.str());
+        }
+    }
+}
+void AsyncConnection::connectCB(const redisAsyncContext * context, int status)
+{
+    if (status != REDIS_OK)
+    {
+        if (context->data)
+        {
+            const AsyncConnection * connection = (AsyncConnection *)context->data;
+            VStringBuffer msg("Redis Plugin : failed to connect to %s:%d - %s", connection->master.str(), connection->port, context->errstr);
+            rtlFail(0, msg.str());
```

Comment:
Do you want failures in the callback?  Will they get fed back to the correct place?